lua/plenary/enum.lua

```diff
@@ -0,0 +1,137 @@
+--- string or a tuple of the form {string, number}. In the former case, the enum
+--- member will take the next available value, while in the latter, the member
+--- will take the string as it's name and the number as it's value. In both
+--- cases, the name must start with a capital letter.
+---
+--- Here is an example:
+---
+--- <pre>
+--- local Enum = require 'plenary.enum'
+--- local myEnum = Enum {
+---     'Foo',          -- Takes value 1
+---     'Bar',          -- Takes value 2
+---     {'Qux', 10},    -- Takes value 10
+---     'Baz',          -- Takes value 11
+--- }
+--- </pre>
+---
+--- In case of name or value clashing, the call will fail. For this reason, it's
+--- best if you define the members in ascending order.
+---@brief ]]
+local Enum = {}
+
+---@class Enum
+
+local function validate_member_name(name)
+  if #name > 0 and name:sub(1, 1):match('%u') then return name end
+  error(name .. ' should start with a capital letter')
+end
+
+--- Creates an enum from the given list-like table, like so:
+--- <pre>
+--- local enum = Enum.make_enum{
+---     'Foo',
+---     'Bar',
+---     {'Qux', 10}
+--- }
+--- </pre>
+--- @return Enum: A new enum
+local function make_enum(tbl)
+  local enum = {}
+
+  local Variant = {}
+  Variant.__index = Variant
+
+  local function newVariant(i)
+    return setmetatable({_id = i}, Variant)
+  end
+
+  -- we don't need __eq because the __eq metamethod will only ever be
+  -- invoked when they both have the same metatable
+
+  function Variant:__lt(o)
+    return self._id < o._id
+  end
+
+  function Variant:__gt(o)
+    return self._id > o._id
+  end
+
+  function Variant:__tostring()
+    return tostring(self._id)
+  end
+
+  function Variant:id()
+    return self._id
+  end
+
+  function Variant:is(other)
+    if getmetatable(other) == Variant then return self._id == other._id end
+    error('is() method must be used with elements from the same enum')
+  end
+
+  local function find_next_idx(enum, i)
+    local newI = i + 1
+    if not enum[newI] then return newI end
+    error('Overlapping index: ' .. tostring(newI))
+  end
+
+  local i = 0
+
+  for _, v in ipairs(tbl) do
+    if type(v) == 'string' then
+      local name = validate_member_name(v)
+      local idx = find_next_idx(enum, i)
+      enum[idx] = name
+      if enum[name] then error('Duplicate enum member name: ' .. name) end
+      enum[name] = newVariant(idx)
+      i = idx
+    elseif type(v) == 'table' and type(v[1]) == 'string' and type(v[2])
+        == 'number' then
+      local name = validate_member_name(v[1])
+      local idx = v[2]
+      if enum[idx] then error('Overlapping index: ' .. tostring(idx)) end
+      enum[idx] = name
+      if enum[name] then error('Duplicate name: ' .. name) end
+      enum[name] = newVariant(idx)
+      i = idx
+    else
+      error('Invalid way to specify an enum variant')
+    end
+  end
+
+  return setmetatable(enum, Enum)
```

Comment:
I just merged `tbl.freeze` so we can do `return tbl.freeze(setmetatable(enum, Enum))`